### PR TITLE
feat(VBadge): port to v3

### PR DESCRIPTION
**Pull request by nekosaur**
## Description

Updated v-badge for v3

## Breaking changes

- ~~Changed `overlap` default value to `true`. In my opinion this looks better for most use cases.~~
- Removed `overlap` prop and made default. For existing style, use the `floating` prop
- Removed `avatar` prop
- Removed theme props
- Removed `left` prop, use `location` instead
- Removed `bottom` prop, use `location` instead
- Removed `mode` prop, use `transition` instead
- Removed `origin` prop, use `transition` instead
- `offset-x` and `offset-y` now offset the original positioning as their name implies instead of absolutely positioning the badge
- `tile` is removed (removal inherited from border-radius composable)
- `value` is now `model-value`

## Markup:

<details>

```vue
<template>
  <v-app class="fill-height">
    <v-container style="height: 100%" class="d-flex align-center">
      <v-row
        align="center"
      >
        <v-col cols="12" class="d-flex justify-space-around">
          <v-badge
            location="top-left"
            content="4"
          >
            <v-btn color="secondary">top left</v-btn>
          </v-badge>

          <v-badge
            location="top-right"
            content="4"
          >
            <v-btn color="secondary">top right</v-btn>
          </v-badge>

          <v-badge
            location="bottom-right"
            content="4"
          >
            <v-btn color="secondary">bottom right</v-btn>
          </v-badge>

          <v-badge
            location="bottom-left"
            content="4"
          >
            <v-btn color="secondary">bottom left</v-btn>
          </v-badge>
        </v-col>

        <v-col cols="12" class="d-flex justify-space-around">
          <v-badge
            bordered
            color="secondary"
            location="bottom-left"
          >
            <template #badge>
              <v-icon icon="mdi-home" />
            </template>
            <v-btn
              color="error"
              elevation="0"
            >
              Lock Account
            </v-btn>
          </v-badge>

          <v-badge
            bordered
            dot
            color="success"
            location="bottom-right"
            offset-x="4"
            offset-y="2"
          >
            <v-avatar>
              <v-img src="https://cdn.vuetifyjs.com/images/lists/2.jpg" />
            </v-avatar>
          </v-badge>

          <v-badge
            bordered
          >
            <template #badge>
              <v-img src="https://cdn.vuetifyjs.com/images/logos/v.png" />
            </template>

            <v-avatar>
              <v-img src="https://cdn.vuetifyjs.com/images/john.png" />
            </v-avatar>
          </v-badge>

          <v-badge
            bordered
          >
            <template #badge>
              <img src="https://cdn.vuetifyjs.com/images/logos/v.png">
            </template>

            <v-avatar>
              <v-img src="https://cdn.vuetifyjs.com/images/john.png" />
            </v-avatar>
          </v-badge>

          <v-badge
            content="2"
            color="secondary"
            :rounded="zxc ? 0 : undefined"
            :model-value="show"
            :aria-label="zxc ? 'foo' : 'bar'"
          >
            <v-btn @click="show = !show">hello</v-btn>
          </v-badge>

          <v-btn @click="zxc = !zxc">toggle</v-btn>
        </v-col>
        <v-col cols="12" class="d-flex justify-space-around">
          <v-badge
            :model-value="hover"
            content="9999+"
            location="bottom right"
            transition="scale-transition"
            floating
          >
            <!-- <v-hover v-model="hover"> -->
            <v-icon
              color="#888888"
              size="x-large"
              @mouseenter="hover = true"
              @mouseleave="hover = false"
            >
              mdi-account
            </v-icon>
            <!-- </v-hover> -->
          </v-badge>

          <v-badge content="9" icon="mdi-close">
            <v-btn icon>
              <v-icon icon="mdi-home" />
            </v-btn>
          </v-badge>

          <div>
            This is an inline badge (maybe should be vertical-align: text-top?)
            <v-badge content="3" inline />
          </div>

          <v-badge content="99" floating>
            This is just some regular old text
          </v-badge>
        </v-col>
      </v-row>
    </v-container>
  </v-app>
</template>

<script>
  import { ref } from 'vue'

  export default {
    setup () {
      const show = ref(true)
      const zxc = ref(true)
      const hover = ref(false)

      return { show, zxc, hover }
    },
  }
</script>
```
</details>

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Improvement/refactoring (non-breaking change that doesn't add any features but makes things better)

## Checklist:

- [x] The PR title is no longer than 64 characters.
- [x] The PR is submitted to the correct branch (`master` for bug fixes and documentation updates, `dev` for new features and backwards compatible changes and `next` for non-backwards compatible changes).
- [x] My code follows the code style of this project.
- [x] I've added relevant changes to the documentation (applies to new features and breaking changes in core library)

### CODE REVIEW

File: packages/vuetify/src/components/VBadge/VBadge.sass [hunk 1]
```diff
@@ -1,56 +1,45 @@
...
     line-height: 1
-    min-width: $badge-min-width
     padding: $badge-padding
     pointer-events: auto
     position: absolute
     text-align: center
     text-indent: 0
-    top: $badge-top
-    transition: $primary-transition
+    // transition: $primary-transition
```
Comment (johnleider) [inline packages/vuetify/src/components/VBadge/VBadge.sass:23]: value should be in variables: `transition: .3s $standard-easing !default`

File: packages/vuetify/src/components/VBadge/VBadge.sass [hunk 2]
```diff
@@ -59,12 +48,11 @@
       .v-avatar
         height: $badge-height !important
```
Comment (johnleider) [inline packages/vuetify/src/components/VBadge/VBadge.sass:49]: do we still need important here?

File: packages/vuetify/src/components/VBadge/VBadge.sass [hunk 3]
```diff
@@ -59,12 +48,11 @@
...
         min-width: 0 !important
-        max-width: $badge-min-width !important
 
   &--bordered
     .v-badge__badge::after
       border-radius: inherit
-      border-width: $badge-bordered-width
+      border-width: $badge-border-width
       border-style: solid
```
Comment (johnleider) [inline packages/vuetify/src/components/VBadge/VBadge.sass:56]: value should be in variables

File: packages/vuetify/src/components/VBadge/VBadge.sass [hunk 4]
```diff
@@ -73,6 +61,7 @@
       right: 0
       top: 0
       transform: scale(1.15)
```
Comment (johnleider) [inline packages/vuetify/src/components/VBadge/VBadge.sass:63]: value should be in variables

File: packages/vuetify/src/components/VBadge/VBadge.tsx [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+      required: false
+    },
+    dot: Boolean,
+    label: {
+      type: String,
+      default: '$vuetify.badge',
+    },
+    offsetX: [Number, String],
+    offsetY: [Number, String],
+    overlap: {
```
Comment (johnleider) [inline packages/vuetify/src/components/VBadge/VBadge.tsx:55]: I feel boolean true props are a better interface to work with. Maybe a new prop named `floating`?

File: packages/vuetify/src/components/VBadge/VBadge.tsx [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+
+  setup (props, ctx) {
+    const { backgroundColorClasses, backgroundColorStyles } = useBackgroundColor(toRef(props, 'color'))
+    const { textColorClasses, textColorStyles } = useTextColor(toRef(props, 'textColor'))
+    const { borderRadiusClasses } = useBorderRadius(props)
+
+    const position = computed(() => {
+      if (props.overlap) return props.dot ? 8 : 12
+      return props.dot ? 2 : 4
+    })
```
Comment (johnleider) [inline packages/vuetify/src/components/VBadge/VBadge.tsx:90]:
```suggestion
    const position = computed(() => props.overlap ? props.dot ? 8 : 12 : props.dot ? 2 : 4)
```

File: packages/vuetify/src/components/VBadge/VBadge.tsx [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+                  }
+                >
+                  { props.dot
+                    ? undefined
+                    : ctx.slots.badge
+                      ? ctx.slots.badge()
+                      : props.icon
+                        ? <VIcon icon={props.icon} />
+                        : props.content
+                  }
```
Comment (johnleider) [inline packages/vuetify/src/components/VBadge/VBadge.tsx:166]:
```suggestion
                  { ctx.slots.badge?.() }

                  { props.icon ? <VIcon icon={props.icon} /> : props.content }
```

File: packages/vuetify/src/components/VBadge/index.ts [hunk 8]
```diff
@@ -1,4 +1,3 @@
 import VBadge from './VBadge'
 
 export { VBadge }
```
Comment (johnleider) [inline packages/vuetify/src/components/VBadge/index.ts:3]:
```suggestion
export { default as VBadge } from './VBadge'
```

File: packages/vuetify/src/composables/transition.ts [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+import { consoleWarn } from '@/util/console'
+
+type TransitionProp = string | boolean | TransitionProps
+
+export const makeTransitionProps = propsFactory({
+  transition: {
+    type: [String, Boolean, Object] as PropType<TransitionProp>,
+    validator: (value: any) => {
+      return value !== true
+    },
```
Comment (johnleider) [inline packages/vuetify/src/composables/transition.ts:16]:
```suggestion
    validator: (value: any) => value !== true,
```

File: packages/vuetify/src/composables/transition.ts [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+  if (transition === true) {
+    consoleWarn('`transition` prop does not accept value `true`')
+    return vnode
+  }
+
+  const transitionProps = typeof transition === 'string' ? { name: transition } : transition
+
+  return h(Transition, transitionProps, {
+    default: () => vnode,
+  })
```
Comment (johnleider) [inline packages/vuetify/src/composables/transition.ts:34]:
```suggestion
  if (typeof transition === 'boolean') {
    transition === true && consoleWarn('`transition` prop does not accept value `true`')

    return vnode
  }

  return h(Transition,
    typeof transition === 'string' ? { name: transition } : transition,
    { default: () => vnode }
  )
```

Comment (johnleider) [reply packages/vuetify/src/composables/transition.ts:34]: 🚲

**Review by johnleider — CHANGES_REQUESTED** [review]
scss/sass imports should match v-banner, v-btn, etc

File: packages/vuetify/src/components/VBadge/VBadge.tsx (hunk 6 above)
Comment (nekosaur) [reply packages/vuetify/src/components/VBadge/VBadge.tsx:90]: much less readable imo

**Comment by johnleider** [comment]
Requesting **Initial** comments from @vuetifyjs/contributors @vuetifyjs/core-team

File: yarn.lock [hunk 11]
```diff
@@ -4653,7 +4653,7 @@
     webpack "^4.0.0"
     webpack-bundle-analyzer "^3.8.0"
     webpack-chain "^6.4.0"
-    webpack-dev-server "3.9.0"
+    webpack-dev-server "^3.11.0"
```
Comment (KaelWD) [inline yarn.lock:4656]:
```suggestion
    webpack-dev-server "3.9.0"
```

File: packages/vuetify/src/components/VBadge/VBadge.tsx [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+                  ] as any} // TODO: Fix this :(
+                  aria-atomic="true"
+                  aria-label="locale string here" // TODO: locale string here
+                  aria-live="polite"
+                  role="status"
+                  { ...badgeAttrs }
+                >
+                  {
+                    props.dot ? undefined :
+                    ctx.slots.badge ? ctx.slots.badge?.() :
```
Comment (KaelWD) [inline packages/vuetify/src/components/VBadge/VBadge.tsx:152]:
```suggestion
                    ctx.slots.badge ? ctx.slots.badge() :
```

File: packages/vuetify/src/components/VBadge/VBadge.tsx [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+        <props.tag
+          class={[
+            'v-badge',
+            {
+              'v-badge--bordered': props.bordered,
+              'v-badge--dot': props.dot,
+              'v-badge--floating': props.floating,
+            },
+          ]}
+          { ...attrs }
```
Comment (KaelWD) [inline packages/vuetify/src/components/VBadge/VBadge.tsx:124]: Why `inheritAttrs: false` if you just doing this anyway?

File: packages/vuetify/src/components/VBadge/VBadge.tsx [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+                    borderRadiusClasses.value,
+                    textColorClasses.value,
+                  ]}
+                  style={[
+                    backgroundColorStyles.value,
+                    locationStyles.value,
+                    textColorStyles.value,
+                  ] as any} // TODO: Fix this :(
+                  aria-atomic="true"
+                  aria-label="locale string here" // TODO: locale string here
```
Comment (KaelWD) [inline packages/vuetify/src/components/VBadge/VBadge.tsx:145]: What would this be?

File: packages/vuetify/src/components/VBadge/VBadge.tsx [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+import { makeBorderRadiusProps, useBorderRadius } from '@/composables/border-radius'
+import { makeTagProps } from '@/composables/tag'
+import { makeTransitionProps, withTransition } from '@/composables/transition'
+import { useBackgroundColor, useTextColor } from '@/composables/color'
+
+// Utilities
+import { computed, defineComponent, toRef } from 'vue'
+import { convertToUnit } from '@/util/helpers'
+
+function extract (obj: Dictionary<unknown>, properties: string[]) {
```
Comment (KaelWD) [inline packages/vuetify/src/components/VBadge/VBadge.tsx:17]: Move to helpers

File: packages/vuetify/src/components/VBadge/VBadge.tsx [hunk 16]
```diff
@@ -0,0 +1,165 @@
...
+
+export default defineComponent({
+  name: 'VBadge',
+
+  inheritAttrs: false,
+
+  props: {
+    ...makeBorderRadiusProps(),
+    ...makeTagProps(),
+    ...makeTransitionProps({ transition: 'scale-rotate-transition' }),
```
Comment (KaelWD) [inline packages/vuetify/src/components/VBadge/VBadge.tsx:40]: Move to the end.

File: packages/vuetify/src/components/VBadge/VBadge.tsx [hunk 17]
```diff
@@ -0,0 +1,165 @@
...
+
+  setup (props, ctx) {
+    const { backgroundColorClasses, backgroundColorStyles } = useBackgroundColor(toRef(props, 'color'))
+    const { borderRadiusClasses } = useBorderRadius(props)
+    const { textColorClasses, textColorStyles } = useTextColor(toRef(props, 'textColor'))
+
+    const position = computed(() => {
+      if (!props.floating) return props.dot ? 8 : 12
+
+      return props.dot ? 2 : 4
```
Comment (KaelWD) [inline packages/vuetify/src/components/VBadge/VBadge.tsx:83]:
```suggestion
      return props.floating
        ? (props.dot ? 2 : 4)
        : (props.dot ? 8 : 12)
```

**Comment by KaelWD** [comment]
> Removed inline prop. I couldn't figure our what the use-case for this prop was?

Something like facebook's new message notification maybe? 
image

File: packages/vuetify/src/components/VBadge/VBadge.tsx [hunk 18]
```diff
@@ -0,0 +1,165 @@
...
+              'v-badge--floating': props.floating,
+            },
+          ]}
+          { ...attrs }
+        >
+          <div class="v-badge__wrapper">
+            { ctx.slots.default?.() }
+
+            {
+              withTransition(
```
Comment (KaelWD) [inline packages/vuetify/src/components/VBadge/VBadge.tsx:130]: How is this different to `maybeTransition`?

Comment (nekosaur) [reply packages/vuetify/src/components/VBadge/VBadge.tsx:130]: sorry, didn't realize we had it

File: packages/vuetify/src/components/VBadge/VBadge.tsx (hunk 14 above)
Comment (nekosaur) [reply packages/vuetify/src/components/VBadge/VBadge.tsx:145]: `t(props.label)`

Comment (KaelWD) [reply packages/vuetify/src/components/VBadge/VBadge.tsx:145]: It's just `Badge`? That seems pretty useless.

**Comment by nekosaur** [comment]
> Something like facebook's new message notification maybe?

Hmm yeah maybe. If so would it make sense to do

```html
<div>
  <div>something that needs inline badge after it</div>
  <v-badge inline content="99" />
</div>
```

Instead of 
```html
<div>
  <v-badge inline content="99">
  	<div>something that needs inline badge after it</div>
  </v-badge>
</div>
```

My initial thoughts are that the first example makes it easier to position the badge in whatever way you need. E.g.

```html
<div class="d-flex align-center justify-space-between">
  <div>something that needs inline badge after it</div>
  <v-badge inline content="99" />
</div>
```

File: packages/vuetify/src/components/VBadge/VBadge.tsx (hunk 14 above)
Comment (nekosaur) [reply packages/vuetify/src/components/VBadge/VBadge.tsx:145]: 🤷 perhaps

**Review by johnleider — APPROVED** [review]

**Comment by johnleider** [comment]
Requesting **Final** comments from @vuetifyjs/core-team

**Review by Mert75 — APPROVED** [review]
Looks good to me 👍

File: packages/api-generator/src/locale/en/v-badge.json [hunk 19]
```diff
@@ -1,16 +1,16 @@
 {
   "props": {
-    "avatar": "Removes badge padding for the use of the `v-avatar` in the **badge** slot.",
     "bordered": "Applies a **2px** by default and **1.5px** border around the badge when using the **dot** property.",
-    "content": "Any content you want injected as text into the badge.",
-    "dot": "Reduce the size of the badge and hide its contents",
-    "icon": "Designates a specific icon used in the badge.",
-    "inline": "Moves the badge to be inline with the wrapping element. Supports the usage of the **left** prop.",
```
Comment (nekosaur) [inline packages/api-generator/src/locale/en/v-badge.json:8]: inline is back

File: packages/api-generator/src/locale/en/v-badge.json [hunk 20]
```diff
@@ -1,16 +1,16 @@
...
+    "dot": "Reduce the size of the badge and hide its contents.",
+    "icon": "Designates a specific icon to render in the badge.",
+    "label": "The **aria-label** used for the badge.",
     "offsetX": "Offset the badge on the x-axis.",
     "offsetY": "Offset the badge on the y-axis.",
     "overlap": "Overlaps the slotted content on top of the component.",
-    "tile": "Removes the component's border-radius."
+    "location": "Changes the location of the badge. Possible values are **top-right**, **bottom-right**, **bottom-left**, **top-left**.",
+    "color": "Defines the background color of the badge.",
+    "textColor": "Defines the text color of the badge."
```
Comment (nekosaur) [inline packages/api-generator/src/locale/en/v-badge.json:13]: missing `max`

File: packages/vuetify/src/components/VApp/VApp.tsx [hunk 21]
```diff
@@ -30,7 +30,9 @@ export default defineComponent({
         ]}
         data-app="true"
       >
-        <div class="v-application__wrap" v-slots={ slots } />
+        <div class="v-application__wrap">
```
Comment (nekosaur) [inline packages/vuetify/src/components/VApp/VApp.tsx:33]: unrelated?

File: packages/vuetify/src/components/VMain/VMain.tsx [hunk 22]
```diff
@@ -27,7 +27,9 @@ export default defineComponent({
           ssrBootStyles.value,
         ]}
       >
-        <div class="v-main__wrap" v-slots={ slots } />
+        <div class="v-main__wrap">
```
Comment (nekosaur) [inline packages/vuetify/src/components/VMain/VMain.tsx:30]: unrelated?

File: packages/vuetify/src/composables/transition.ts [hunk 23]
```diff
@@ -0,0 +1,13 @@
+// Utilities
+import propsFactory from '@/util/propsFactory'
+
+// Types
+import type { PropType } from 'vue'
+
+export const makeTransitionProps = propsFactory({
+  transition: {
+    type: [Boolean, String] as PropType<string | false>,
```
Comment (nekosaur) [inline packages/vuetify/src/composables/transition.ts:9]: this does not have the object variant that I added. but it can be re-implemented in a separate PR

File: packages/vuetify/src/components/VBadge/__tests__/VBadge.spec.tsx [hunk 24]
```diff
@@ -0,0 +1,54 @@
...
+    [{ floating: true }],
+    [{ icon: 'foo' }], // badge slot overrides icon
+    [{ inline: true }],
+    [{ location: 'bottom-left' }],
+    [{ location: undefined }],
+    [{ location: null }],
+    [{ content: '10', max: 9 }],
+    [{ content: '10', max: 11 }],
+    [{ content: 'foo', max: 9 }],
+  ])('should render component and match snapshot with badge slot', (props: any) => {
```
Comment (nekosaur) [inline packages/vuetify/src/components/VBadge/__tests__/VBadge.spec.tsx:33]: if possible please improve this title so that the snapshots are easier to distinguish

File: packages/vuetify/src/components/VBadge/__tests__/VBadge.spec.tsx [hunk 25]
```diff
@@ -0,0 +1,54 @@
...
+    })
+
+    expect(wrapper.html()).toMatchSnapshot()
+  })
+
+  it.each([
+    [{ content: 'foo' }],
+    [{ content: 'foo', icon: 'bar' }],
+    [{ icon: 'foo' }],
+  ])('should render component and match snapshot without badge slot', (props: any) => {
```
Comment (nekosaur) [inline packages/vuetify/src/components/VBadge/__tests__/VBadge.spec.tsx:49]: same here

File: packages/vuetify/src/components/VBadge/VBadge.tsx [hunk 26]
```diff
@@ -0,0 +1,163 @@
...
+      return {
+        bottom: 'auto',
+        left: 'auto',
+        right: 'auto',
+        top: 'auto',
+        ...(!props.inline ? {
+          [horizontal === 'left' ? 'right' : 'left']: calculatePosition(props.offsetX),
+          [vertical === 'top' ? 'bottom' : 'top']: calculatePosition(props.offsetY),
+        } : {}),
+      }
```
Comment (nekosaur) [inline packages/vuetify/src/components/VBadge/VBadge.tsx:91]:
```suggestion
      const styles = {
        bottom: 'auto',
        left: 'auto',
        right: 'auto',
        top: 'auto'
      }
      
      if (!props.inline) {
        styles[horizontal === 'left' ? 'right' : 'left'] = calculatePosition(props.offsetX)
        styles[vertical === 'top' ? 'bottom' : 'top'] = calculatePosition(props.offsetY)
      }
      
      return styles
```

?

**Comment by nekosaur** [comment]
Haven't actually tried the code yet

File: packages/vuetify/src/components/VBadge/_variables.scss [hunk 27]
```diff
@@ -1,21 +1,23 @@
-@import '../../styles/styles.sass';
-
+// Defaults
+$badge-border-color: rgb(var(--v-theme-background)) !default;
 $badge-border-radius: 10px !default;
-$badge-bordered-width: 2px !default;
-$badge-color: map-get($shades, 'white') !default;
+$badge-border-style: solid !default;
+$badge-border-transform: scale(1.20) !default;
+$badge-border-width: 2px !default !default;
```
Comment (nekosaur) [inline packages/vuetify/src/components/VBadge/_variables.scss:6]: double default?

**Review by KaelWD — APPROVED** [review]

**Review by KaelWD — APPROVED** [review]
